### PR TITLE
Delete `y.grad` on `y.backward()`

### PR DESCRIPTION
**Pull request by toslunar**

### CODE REVIEW

File: chainer/_backprop_utils.py [hunk 1]
```diff
@@ -53,7 +53,7 @@ def get_as_list(self, node):
             if self._load_if_new and node.creator_node is None:
                 node._check_old_style_gradient()
                 # accumulate the gradient only if the node is a leaf
-                grads[node] = _pure(node.grad_var)
+                grads[node] = _pure(node._pop_grad_var_if_available())
```
Comment (niboshi) [inline chainer/_backprop_utils.py]: `GradTable` is no longer a simple data holder but it describes the algorithm, because it modifies the external data structure.
This behavior is not clearly expressed and therefore difficult to understand the code, IMO.
Perhaps the method name `get_as_list` should be something that expresses that behavior.
Or perhaps the class name `GradTable` should be something that does not imply a simple data holder.

File: chainer/variable.py [hunk 2]
```diff
@@ -313,6 +313,16 @@ def _set_grad_var_if_available(self, g):
         if var is not None:
             var._set_grad_var_without_check(g)
 
+    def _pop_grad_var_if_available(self):
```
Comment (niboshi) [inline chainer/variable.py]: How about making this method a member of `GradTable`?
This set of operations does not seem to make sense for `VariableNode`.

(You can simply use `VariableNode.variable_or_none()` and `Variable.grad_var.setter`.)

Comment (niboshi) [inline chainer/variable.py]: The meaning of `_if_available` is not clear for me.
Perhaps `_if_alive` is better, or perhaps it can simply be removed.

File: chainer/variable.py [hunk 3]
```diff
@@ -1562,19 +1587,11 @@ def add_cand(cand):
...
-    for y_var in outputs:
-        # TODO(sonots): Implement for ChainerX
-        if y_var.xp is chainerx:
-            raise NotImplementedError()
-
-        y = y_var.node
-        root_nodes.add(y)
-        grads[y] = y_var.grad_var
+    for y, gy in outputs:
+        grads.get_as_list(y).append(gy)
```
Comment (niboshi) [inline chainer/variable.py]: How about creating a new method in `GradTable` for this?
This line looks unusual and perhaps surprising that `append` affects `grads`.

File: chainer/variable.py [hunk 4]
```diff
@@ -1631,11 +1648,12 @@ def add_cand(cand):
                 hook.backward_postprocess(
                     func, tuple(in_data), tuple(out_grad_array))
 
-        for y, gy in six.moves.zip(outputs, out_grad):
-            if y is not None and y not in root_nodes:
-                y._set_grad_var_if_available(
-                    gy if retain_grad else None)
-        del gy, out_grad  # to reduce memory usage
+        if retain_grad:
```
Comment (niboshi) [inline chainer/variable.py]: How about writing a block comment?

File: tests/chainer_tests/test_variable.py [hunk 5]
```diff
@@ -371,28 +371,37 @@ def test_label_gpu(self):
         self.check_label(self.label, cuda.to_gpu(self.c))
 
     def check_backward(self, inputs, intermediates, outputs, retain_grad):
-        for o in outputs:
-            o.backward(retain_grad)
+        # This test assumes that `outputs` do not depend each other
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: How about writing a brief summary of this test?
(I think it's not easily guessed.)

File: tests/chainer_tests/test_variable.py [hunk 6]
```diff
@@ -2605,6 +2617,49 @@ def test_str_gpu(self):
...
+        ]),
+
+
+def exp_pair(x):
+    return ExpPair().apply((x,))
+
+
+class TestBackwardRetainedOutput(unittest.TestCase):
+
+    def test_backward_cpu_retain_grad(self):
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Ditto. Please write the objective of this test.

File: tests/chainer_tests/test_variable.py [hunk 7]
```diff
@@ -2605,6 +2617,49 @@ def test_str_gpu(self):
...
+        del y0
+        y1.grad = np.array([3, 4], np.float32)
+        y1.backward(retain_grad=True)
+        gx_expected = y1.array * y1.grad
+        testing.assert_allclose(x.grad, gx_expected)
+
+
+class TestBackwardV5Compat1(unittest.TestCase):
+
+    def test_backward(self):
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Ditto. Please write the objective of this test.

File: chainer/_backprop_utils.py (hunk 1 above)
Comment (toslunar) [reply chainer/_backprop_utils.py]: - For the line, I'll move the logic from `get_as_list` to `_backprop_to_all`.
- `GradTable.pop` is not simple because of an issue that #5722 will fix.

File: tests/chainer_tests/test_variable.py [hunk 8]
```diff
@@ -2605,6 +2630,59 @@ def test_str_gpu(self):
...
+            if g is not None
+        ]),
+
+
+def exp_pair(x):
+    return ExpPair().apply((x,))
+
+
+class TestBackwardDelOutput(unittest.TestCase):
+    # Variable.backward may have function nodes with deleted outputs.
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: How about this:
>`# Variable.backward may be applied to function nodes with deleted outputs.`

File: tests/chainer_tests/test_variable.py [hunk 9]
```diff
@@ -2605,6 +2630,59 @@ def test_str_gpu(self):
...
+
+    def check_backward_cpu(self, retain_grad):
+        x = chainer.Variable(np.array([1, 2], np.float32))
+        y0, y1 = exp_pair(x)
+        del y0
+        y1.grad = np.array([3, 4], np.float32)
+        y1.backward(retain_grad=True)
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Why `retain_grad=True`?

File: tests/chainer_tests/test_variable.py [hunk 10]
```diff
@@ -2605,6 +2630,59 @@ def test_str_gpu(self):
...
+        gx_expected = y1.array * y1.grad
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: It's better to calculate this before `backward`.

File: tests/chainer_tests/test_variable.py [hunk 11]
```diff
@@ -2605,6 +2630,59 @@ def test_str_gpu(self):
...
+            for g, exp in zip(grad_outputs, self.get_retained_outputs())
+            if g is not None
+        ]),
+
+
+def exp_pair(x):
+    return ExpPair().apply((x,))
+
+
+class TestBackwardDelOutput(unittest.TestCase):
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Is this test related to this PR?

File: tests/chainer_tests/test_variable.py [hunk 12]
```diff
@@ -2605,6 +2630,59 @@ def test_str_gpu(self):
...
+        testing.assert_allclose(x.grad, gx_expected)
+
+    def test_backward_cpu(self):
+        self.check_backward_cpu(False)
+
+    def test_backward_cpu_retain_grad(self):
+        self.check_backward_cpu(True)
+
+
+class TestBackwardV5Compat(unittest.TestCase):
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Ditto.
Is this test related to this PR?

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM
